### ISSUE TITLE
systemd-nspawn: --network-interface does not move the interface back to the host namespace

### ISSUE THREAD

**Issue by mephi42** [systemd/systemd#4330]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 219
> Judging by the code, bug is still present in master: interfaces are moved into container namespace here https://github.com/systemd/systemd/blob/master/src/nspawn/nspawn.c#L4056 , but there is no code to move them back.
### Used distribution

> CentOS 7
### In case of bug report: Expected behaviour you didn't see

https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html
--network-interface=
... When the container terminates, it is moved back to the host namespace ...
So:

```
host# ip tuntap add tap0 mode tap
host# systemd-nspawn --network-interface=tap0
container# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
14: tap0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT qlen 500
    link/ether 9e:a5:79:2d:42:f0 brd ff:ff:ff:ff:ff:ff
container# exit
```

Expected:

```
host# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
...
14: tap0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT qlen 500
    link/ether xx:xx:xx:xx:xx:xx brd ff:ff:ff:ff:ff:ff
```
### In case of bug report: Unexpected behaviour you saw

```
host# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
...
(tap0 is missing)
```
### In case of bug report: Steps to reproduce the problem

See above

**Comment by lucaswerkmeister** [comment]
I can reproduce this issue with the above steps on Arch Linux, systemd version 231. (Edit: `uname -a` says Linux 4.7.6-1-ARCH.)

**Comment by poettering** [comment]
Well, it's the kernel's job to move them back really. If you run the same with a currentl nspawn on a current kernel, do they still get lost?

**Comment by mephi42** [comment]
Retried on archlinux, got the same result:

```
host# uname -a
Linux localhost 4.8.1-1-ARCH #1 SMP PREEMPT Fri Oct 7 19:14:45 CEST 2016 x86_64 GNU/Linux
host# systemctl --version
systemd 231
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN
host# ip tuntap add tap0 mode tap
host# systemd-nspawn --network-interface=tap0
container# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
3: tap0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether ea:e9:3d:47:4a:21 brd ff:ff:ff:ff:ff:ff
container# exit
host# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:61:7b:8e brd ff:ff:ff:ff:ff:ff
```

**Comment by poettering** [comment]
hmm, so here's an idea. maybe the kernel behaviour is like this: if the device is of a type that can be deleted (i.e. one where "ip link delete" can be used on, i.e. all "virtual" ones, like tunnels, tap, and so on), it will delete it when the namespace goes away. For all others, where they cannot be deleted (i.e. "physical" ones, such as your actual hardware driver's "eth0"), it moves the iface back into host namespace.

Such behaviour would make a ton of sense and explain what you are seeing. If this is the case, then we should probably document this correctly in the man page.

On your system, if you use --network-interface=ens3 (i.e. use it on a physical one, instead of a virtual one) does the interface come back properly onto the host when the container dies?

**Comment by mephi42** [comment]
On archlinux ens3 is moved back to the host.

However, the original issue for me was that CentOS 7 did not restore physical adapters (I used tap only for the minimal repro) and I had to reboot to get them back. The difference with archlinux's ens3 was that there links were down. I would like to double check this tomorrow.

**Comment by poettering** [comment]
are you saying that on centos 7 you make the physical adapters completely disappear? If so, that'd be a a kernel bug that should be fixed in the centos 7 kernel...

**Comment by evverx** [comment]
> If this is the case, then we should probably document this correctly in the man page

Well, we can move back `tap0`:

``` sh
# move tap0 to the container

# "save" the network namespace
touch inner-child-net-ns
mount --bind /proc/<inner-child-pid>/ns/net  inner-child-net-ns

# stop container

# restore "tap0"
nsenter --net=inner-child-net-ns ip link set tap0 netns $$

# destroy
umount inner-child-net-ns
```

But does it make sense?

**Comment by poettering** [comment]
I'd much rather leave clean-ups for all of this as much as possible to the kernel, so that SIGKILLing nspawn will leave things in the same clean state as terminating it cleanly.

**Comment by mephi42** [comment]
Yes, I can make physical adapters on CentOS 7 disappear. I could not reproduce it with direct `systemd-nspawn` invocation, but it is reproducible with `machinectl`. Here is my setup:
- Add `--network-interface=em4` to `/usr/lib/systemd/system/systemd-nspawn@.service`
- `machinectl start foobar`
- Check the network namespace:

```
# ps -ef
root      5320     1  0 07:57 ?        00:00:00 /usr/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth --network-interface=em4 --machine=foobar
root      5323  5320  0 07:57 ?        00:00:00 /usr/lib/systemd/systemd

# ls -l /proc/5323/ns/net
lrwxrwxrwx. 1 root root 0 Oct 12 07:58 /proc/5323/ns/net -> net:[4026532586]

# find /proc -name net -not -type d | xargs ls -l | grep 4026532586 | wc -l
40
```

A bunch of processes inside the container use it.
- Check `em4` inside the container:

```
# systemd-run -t -M foobar bash 
# ip link show em4
10: em4: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT qlen 1000
    link/ether 40:f2:e9:0a:09:2f brd ff:ff:ff:ff:ff:ff
```

It appears to have been moved successfully.
- `machinectl terminate foobar`
- Check `em4` on the host:

```
# ip link show em4
Device "em4" does not exist.
```

It is not back.
- Check the network namespace:

```
# find /proc -name net -not -type d | xargs ls -l | grep 4026532586 | wc -l
0
```

The namespace seems to be gone.

**Comment by evverx** [comment]
I can't reproduce.

```
$ uname -r
3.10.0-327.36.2.el7.x86_64

$ sudo ethtool --driver eth0
driver: e1000
version: 7.3.21-k8-NAPI
...
```

> The namespace seems to be gone

Did you check the driver?

``` sh
$ ip link show eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT qlen 1000
    link/ether 52:54:00:9f:bd:fd brd ff:ff:ff:ff:ff:ff

$ sudo ethtool --driver eth0
driver: e1000
version: 7.3.21-k8-NAPI
...

$ lsmod | grep e1000
e1000                 149323  0

$ sudo modprobe -r e1000
$ lsmod | grep e1000 # prints nothing
$ ip link show eth0
Device "eth0" does not exist.

$ sudo modprobe e1000
$ ip link show eth0
3: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT qlen 1000
    link/ether 52:54:00:9f:bd:fd brd ff:ff:ff:ff:ff:ff
...
```

Can you reproduce this issue with a trivial container (the busybox idle init or so)?

Anyway, I think this is not our bug. We rely on the kernel behaviour here.

**Comment by poettering** [comment]
If you indeed can make network devices disappear that way, that's really a kernel bug. Any chance you can file a but about that to CentOS?

Closing this here, as it appears to be a bug in the kernel, not nspawn

**Comment by DaanDeMeyer** [comment]
@poettering Would a patch making sure that virtual devices are moved back to the host network namespace be accepted? I ran into a problem with `--network-veth-extra` where any configuration done on the created `veth` link is lost when the container reboots. More specifically, I added the veth created by nspawn to a bridge and configured vlans on it. However, all this configuration is lost when the container reboots on its own as nspawn deletes and recreates the veth link on reboot. I was looking to get around this by creating the veth link myself and passing it to the container with `--network-interface`. I was hoping that when the container rebooted, nspawn would move the interfaces back to the host namespace on container shutdown and move them back into the container when restarting the container, keeping all the applied configuration. Unfortunately, the veth links are deleted on reboot (even though they are documented to be moved back to the host namespace).

If I manually instructed the container to reboot, I could apply the configuration again myself. However, sometimes the container reboots on its own without my knowledge which means I cannot re-apply the configuration to the veth links. Having `--network-interface` move the container side of the veth link back to the host would be really useful in this scenario so the configuration on the host side of the veth link isn't lost.
